extensions/rvadapter/src/main/java/com/google/android/agera/rvadapter/RepositoryPresenterCompilerStates.java

```diff
@@ -53,12 +56,60 @@
   }
 
   /**
+   * Compiler state to optionally enable fine-grained events using {@link DiffUtil}.
+   */
+  interface RPDiff<TVal, TRet> {
+
+    /**
+     * Specifies a key extractor {@link Function} to enable generating fine-grained events using
+     * {@link DiffUtil}. Two item objects are {@linkplain Callback#areItemsTheSame the same item} if
+     * their keys are {@linkplain Object#equals equal}; and they have {@linkplain
+     * Callback#areContentsTheSame the same content} if they themselves are {@linkplain
+     * Object#equals equal}. However, if the presenter is given the same instance of collection
+     * object to generate the fine-grained events, then the presenter will consider it caused by an
+     * update from an {@linkplain Builder#addAdditionalObservable additional observable}, and send a
+     * blanket item change event covering all positions.
+     *
+     * <p>Note that {@link DiffUtil} may not be suitable for all situations due to the computation
+     * complexity and that <i>the computation is likely done from the main thread</i>. If client
+     * code has better alternatives, it should directly subclass {@link RepositoryPresenter} and
+     * provide a custom implementation in {@link RepositoryPresenter#getUpdates}.
+     *
+     * @param keyForItem A function from individual item objects to objects uniquely identifying the
```

Comment:
key can well be an object, like a string, or a custom ID class. itemIdForItem needs to do one more step of translation from that key to a long. So diffing with a key can be quicker than diffing with the item ID.

with fine-grained events, you theoretically don't need the item IDs anymore if the only thing you worry is animations. however, you can still turn on stable IDs to make RV slightly smarter when it comes to reusing views (it'll favor the view previously bound to an item with the same stable ID)